### ISSUE TITLE
`configure_file()` needs an option to verify all `configuration_data()` has been consumed

### ISSUE THREAD

**Issue by krader1961** [mesonbuild/meson#2978]
I'm working on bringing the [Korn shell project](https://github.com/att/ast/) back to life along with a few others like @siteshwar from Red Hat. We switched the project from the legacy Nmake build system to Meson and are extremely happy with that decision and the behavior of Meson (this has sped up the build by an order of magnitude).

Recently I implemented a change to switch from passing all the platform feature symbols via compiler `-Dsomething` flags to use the `configure_file()` mechanism. Because the legacy code has a huge number of platform feature symbols (approximately 600) I wanted to enumerate all of them in the input file for `configure_file()` to help us understand which symbols were not being defined by the feature detection logic. For example:

```
#mesondefine _hdr_wchar
#mesondefine _hdr_wctype

#mesondefine _lib_BSDsetpgrp
#mesondefine _lib_MD4Init
```

I found that I had omitted several symbols from the template that were in the `configuration_data()` object. But that wasn't immediately obvious because Meson silently ignores any symbols in the `configuration_data()` object which are not consumed by the template. It would make debugging such problems easier if there were a way to tell `configure_file()` that it should fail if any of the symbols were not used by the template. Personally I think that should have been the default behavior but obviously we can't make that the default now as it isn't backward compatible.

I'd like to add something like a `strict_subset : true` optional argument to indicate the template must consume all the config data symbols. A `strict : true` arg to indicate that the config object and the template must contain the same set of symbols might also be useful. Once there is agreement that a) this feature is useful enough to implement, and b) there is agreement about the API I'm happy to do the work to implement it. This issue is primarily to see if there is agreement on items "a" and "b".

**Comment by nirbheek** [comment]
We previously [started warning](https://github.com/mesonbuild/meson/issues/2090) when symbols in the configure file weren't in the configuration_data().

However, do note that you don't need the input file at all if you want to translate all the defines in the configuration_data() object to a header. Omit the `input:` param and the output will have `#define` lines.

There is no need to maintain two copies of the list of symbols. Just put it all in the build files.

**Comment by krader1961** [comment]
Issue #2090 is a different use case than I am requesting. The AT&T AST (i.e., Korn shell project) use case I outlined above may be sufficiently atypical that it doesn't warrant explicit support.


Our `configure_file()` input file contains lines other than `#mesondefine SOME_SYMBOL`. We would have to add another top level config file if we invoked `configure_file()` without an `input:'some_file'` argument in order to generate a config file that contained every `configuration_data()` symbol.

We've also extracted all the symbols tested by `#if` from the source code into that configure input file. The problem is that in translating the legacy feature detection tests we've made the mistake, at least once, where we misspelled the symbol name when adding it to the `configuration_data()` object passed to `configure_file()`. Detecting that type of mistake would be much easier if we could tell `configure_file()` to at least warn, if not hard error, when a given config symbol is not used in a config file substitution.

**Comment by nirbheek** [comment]
> Issue #2090 is a different use case than I am requesting

Yes, I was merely commenting to show some context. Here's another only-slightly-related issue: https://github.com/mesonbuild/meson/issues/1577

So, no one has asked for substitution to consume all values in a configuration_data object, but I think people will find it useful to have a kwarg like `strict: true`.

**Comment by nirbheek** [comment]
If you'd like to work on implementing this, you can take a look at:

mesonlib.py:do_replacement()
mesonlib.py:do_conf_file()
interpreter.py:func_configure_file()

It should not be difficult to understand the code and add this feature.

**Comment by mitre88** [comment]
I'll implement this as `strict: true` on `configure_file()`, matching @nirbheek's suggestion.

When `strict` is set, configuration fails if any `configuration_data()` key is unused by the template (`@KEY@`, `#mesondefine`, and the cmake equivalents). Default stays `false`.
